BraintreeCore/src/main/java/com/braintreepayments/api/ConfigurationLoader.kt

```diff
@@ -0,0 +1,81 @@
+package com.braintreepayments.api
+
+import android.content.Context
+import android.net.Uri
+import android.util.Base64
+import com.braintreepayments.api.Configuration.Companion.fromJson
+import org.json.JSONException
+
+internal class ConfigurationLoader internal constructor(
+    private val httpClient: BraintreeHttpClient,
+    private val configurationCache: ConfigurationCache
+) {
+    constructor(context: Context, httpClient: BraintreeHttpClient) : this(
+        httpClient, ConfigurationCache.getInstance(context)
+    )
+
+    fun loadConfiguration(authorization: Authorization, callback: ConfigurationLoaderCallback) {
+        if (authorization is InvalidAuthorization) {
+            val message = authorization.errorMessage
+            callback.onResult(null, BraintreeException(message))
+            return
+        }
+        val configUrl = Uri.parse(authorization.configUrl)
+            .buildUpon()
+            .appendQueryParameter("configVersion", "3")
+            .build()
+            .toString()
+        val cachedConfig = getCachedConfiguration(authorization, configUrl)
+
+        cachedConfig?.let {
+            callback.onResult(cachedConfig, null)
+        } ?: run {
+            httpClient[configUrl, null, authorization, HttpClient.RETRY_MAX_3_TIMES, HttpResponseCallback { responseBody, httpError ->
+                responseBody?.let {
+                    try {
+                        val configuration = fromJson(it)
+                        saveConfigurationToCache(configuration, authorization, configUrl)
+                        callback.onResult(configuration, null)
+                    } catch (jsonException: JSONException) {
+                        callback.onResult(null, jsonException)
+                    }
+                } ?: run {
+                    val errorMessageFormat = "Request for configuration has failed: %s"
+                    val errorMessage = String.format(errorMessageFormat, httpError!!.message)
+                    val configurationException = ConfigurationException(errorMessage, httpError)
```

Comment:
```suggestion
                } ?: httpError?.let {
                    val errorMessage = "Request for configuration has failed: ${it.message}"
                    val configurationException = ConfigurationException(errorMessage, httpError)
```

Reply:
i updated but im having second thoughts in general about using `let` to replace every `if != null`. Sometimes it really affects readeability. Good 'ol `if` is sometimes the way to go